Code/GraphMol/FileParsers/CDXMLParser.cpp

```diff
@@ -495,28 +496,193 @@ bool parse_fragment(RWMol &mol, ptree &frag,
   return !skip_fragment;
 }
 
-void set_reaction_data(std::string type, std::string prop, SchemeInfo &scheme,
-                       const std::vector<unsigned int> &frag_ids,
-                       const std::map<unsigned int, size_t> &fragments,
-                       const std::vector<std::unique_ptr<RWMol>> &mols) {
+void set_reaction_data(
+    std::string type, std::string prop, SchemeInfo &scheme,
+    const std::vector<unsigned int> &frag_ids,
+    const std::map<unsigned int, size_t> &fragments,
+    std::map<unsigned int, std::vector<int>> &grouped_fragments,
+    const std::vector<std::unique_ptr<RWMol>> &mols) {
   unsigned int reagent_idx = 0;
   for (auto idx : frag_ids) {
-    auto iter = fragments.find(idx);
-    if (iter == fragments.end()) {
+    auto iter = grouped_fragments.find(idx);
+    if (iter == grouped_fragments.end()) {
       BOOST_LOG(rdWarningLog)
           << "CDXMLParser: Schema " << scheme.scheme_id << " step "
-          << scheme.step_id << " " << type << " fragment " << idx
+          << scheme.step_id << " " << type << " reaction fragment " << idx
           << " not found in document." << std::endl;
       continue;
     }
-    if (iter->second >= mols.size()) {
+    /*if (iter->second >= mols.size()) {
       // shouldn't get here
       continue;
+    }*/
+    for (auto reaction_fragment_id : iter->second) {
+      auto fragment = fragments.find(reaction_fragment_id);
+      if (fragment == fragments.end()) {
+        BOOST_LOG(rdWarningLog)
+            << "CDXMLParser: Schema " << scheme.scheme_id << " step "
+            << scheme.step_id << " " << type << " fragment " << idx
+            << " not found in document." << std::endl;
+        continue;
+      }
+      auto &mol = mols[fragment->second];
+      mol->setProp(CDX_SCHEME_ID, scheme.scheme_id);
+      mol->setProp(CDX_STEP_ID, scheme.step_id);
+      mol->setProp(prop, reagent_idx);
+    }
+    reagent_idx += 1;
+  }
+}
+
+template <class T>
+void visit_children(T &node, std::map<unsigned int, Atom *> &ids,
```

Comment:
The parsing of fragments needed to be moved to a recursive function since they may be embedded further in the documentation.


Additionally, a grouped_fragments map is included to group fragments together for the purposes of reactions.

Ungrouped fragments will end up as vectors of size 1 in the grouped_fragement list.  The reaction schemes in the CDXML docs appear to use the fragment id for ungrouped fragments and the grouped id for grouped fragments, so the grouped_fragments holds both for ease of bookkeeping.


After haven written this, I'll add it to the docs in the code.